mlir/test/lib/Conversion/FuncToLLVM/TestConvertFuncOp.cpp

```diff
@@ -0,0 +1,83 @@
+//===- TestConvertFuncOp.cpp - Test LLVM Conversion of Func FuncOp --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TestDialect.h"
+
+#include "mlir/Conversion/FuncToLLVM/ConvertFuncToLLVM.h"
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+
+using namespace mlir;
+
+namespace {
+
+/// Test helper Conversion Pattern to directly call `convertFuncOpToLLVMFuncOp`
+/// to verify this utility function includes all functionalities of conversion
+struct FuncOpConversion : public ConvertOpToLLVMPattern<func::FuncOp> {
+  FuncOpConversion(const LLVMTypeConverter &converter)
+      : ConvertOpToLLVMPattern(converter) {}
+
+  LogicalResult
+  matchAndRewrite(func::FuncOp funcOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    FailureOr<LLVM::LLVMFuncOp> newFuncOp = mlir::convertFuncOpToLLVMFuncOp(
+        cast<FunctionOpInterface>(funcOp.getOperation()), rewriter,
+        *getTypeConverter());
+    if (failed(newFuncOp))
+      return rewriter.notifyMatchFailure(funcOp, "Could not convert funcop");
+
+    rewriter.eraseOp(funcOp);
+    return success();
+  }
+};
+
+struct TestConvertFuncOp
+    : public PassWrapper<TestConvertFuncOp, OperationPass<ModuleOp>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TestConvertFuncOp)
+
+  void getDependentDialects(DialectRegistry &registry) const final {
+    registry.insert<LLVM::LLVMDialect>();
+  }
+
+  [[nodiscard]] StringRef getArgument() const final {
+    return "test-convert-func-op";
+  }
+
+  [[nodiscard]] StringRef getDescription() const final {
+    return "Tests conversion of `func.func` to `llvm.func` for different "
+           "attributes";
+  }
+
+  void runOnOperation() override {
+    MLIRContext *ctx = &getContext();
+
+    LowerToLLVMOptions options(ctx);
+    // Populate type conversions.
+    LLVMTypeConverter typeConverter(ctx, options);
+
+    RewritePatternSet patterns(ctx);
+    patterns.add<FuncOpConversion>(ctx);
+
+    populateFuncToLLVMConversionPatterns(typeConverter, patterns, nullptr);
```

Comment:
Wouldn't this add the regular func conversion pattern in addition to the one added above? So we don't actually know what we are testing?

Reply:
Thanks a lot for suggestion. I actually started with RewriterBase &. But RewriterBase doesn't have inlineRegionBefore and convertRegionTypes. I prefer RewriterBase & if it works. Do you have suggestion on it?

```c++
  rewriter.inlineRegionBefore(funcOp.getFunctionBody(), newFuncOp.getBody(),
                              newFuncOp.end());
  if (failed(rewriter.convertRegionTypes(&newFuncOp.getBody(), converter,
                                         &result))) {
    return rewriter.notifyMatchFailure(funcOp,
                                       "region types conversion failed");
  }
```

Reply:
> Wouldn't this add the regular func conversion pattern in addition to the one added above? So we don't actually know what we are testing?

Yes, this adds regular func conversion patterns. I created a test only have "llvm.emit_c_interface" to only test it is handled by this function.

Reply:
I don't see why we need the regular func conversion pattern here. It should be enough to exercise the new pattern, and we want tests to be as minimal as possible.

Reply:
Ah, thanks for explanation. I think I misunderstood your feedback. What I need to do is to delete this line right?

I just deleted this line and replaced it with pattern conversion for return op to llvm ( I need this to convert full func op ). Please let me know if this addressed the comment. Thanks in advance!

Reply:
Technically we don't even need to use a pattern, just a walk() should do right?

Reply:
For this one, I have to use pattern as `convertFuncOpToLLVMFuncOp` requires pattern rewriter which I can't create outside of pattern. For `ReturnOpConversion`, I choose to use pattern as I already have pattern rewriter setup in place for funcOp.

```c++
patterns.add<FuncOpConversion>(ctx);
```

Reply:
@ftynse if got a chance, can you review latest changes are good to address feedbacks?
* deleting general pattern and replace it with return op -> llvm.return pattern.

```c++
    populateFuncToLLVMConversionPatterns(typeConverter, patterns, nullptr);
```